[Swift] Spooling protocol extension

wendigo opened the pull request:
Fixes https://github.com/trinodb/trino/issues/22662

TODO:

- [x] Squash all fixups together
- [x] Add @nineinchnick as co-author
- [x] Update https://github.com/trinodb/trino/issues/22662 with the design changes

<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description



<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

wendigo commented:
Rebased

wendigo commented on client/trino-client/src/main/java/io/trino/client/JsonCodec.java:
```diff
@@ -65,6 +66,11 @@ public static <T> JsonCodec<T> jsonCodec(Class<T> type)
 
+    public static <T> JsonCodec<T> jsonCodec(Class<T> type, Module... extraModules)
+    {
+        return new JsonCodec<>(OBJECT_MAPPER_SUPPLIER.get().registerModules(extraModules), type);
```
Yeah, makes sense and some changes can be reverted.

wendigo commented on client/trino-client/src/main/java/io/trino/client/QueryDataClientJacksonModule.java:
```diff
@@ -0,0 +1,66 @@
+
+    public QueryDataClientJacksonModule()
+    {
+        super(QueryDataClientJacksonModule.class.getSimpleName(), Version.unknownVersion());
```
not in the super call:

<img width="707" alt="Screenshot 2024-09-02 at 14 39 39" src="https://github.com/user-attachments/assets/b5648e16-bd9b-42ee-a782-2d6b4d4f595a">

wendigo commented on client/trino-client/src/main/java/io/trino/client/QueryResults.java:
```diff
@@ -52,25 +52,25 @@ public QueryResults(
+        this.updateCount = updateCount;
     }
 
     public QueryResults(
```
Correct. Dropped

wendigo commented on client/trino-client/src/main/java/io/trino/client/StatementClientV1.java:
```diff
@@ -100,10 +103,21 @@ class StatementClientV1
     private final Optional<String> originalUser;
     private final String clientCapabilities;
     private final boolean compressionDisabled;
+    private final JsonCodec<QueryResults> jsonCodec;
```
Done.

wendigo commented on client/trino-client/src/main/java/io/trino/client/RawQueryData.java:
```diff
@@ -0,0 +1,53 @@
+        return new RawQueryData(values);
+    }
+
+    // JSON encoded looses type information. In order for it to be usable, we need to fix types
```
Done.

wendigo commented on client/trino-client/src/main/java/io/trino/client/RawQueryData.java:
```diff
@@ -0,0 +1,53 @@
+
+/**
+ * Class represents QueryData serialized to JSON array of arrays of objects.
+ * It has custom handling and representation in the QueryDataJacksonModule.
```
Done.

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/SegmentLoader.java:
```diff
@@ -0,0 +1,162 @@
+
+    @Override
+    public void close()
+            throws Exception
```
Done

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/DataAttribute.java:
```diff
@@ -0,0 +1,109 @@
+    // Number of rows in the segment
+    ROWS_COUNT("rowsCount", Long.class),
+    // Actual size of the segment in bytes (uncompressed or compressed)
+    BYTE_SIZE("byteSize", Integer.class),
```
Makes sense

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/SegmentLoader.java:
```diff
@@ -0,0 +1,162 @@
+            return delegatingInputStream(response, requireNonNull(body, "response body is null").source().inputStream(), segmentUri);
+        }
+
+        throw new IOException("Could not open segment for streaming " + response.code() + " " + response.message());
```
done

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/SegmentLoader.java:
```diff
@@ -0,0 +1,162 @@
+        ResponseBody body = response.body();
+
+        if (response.isSuccessful()) {
+            return delegatingInputStream(response, requireNonNull(body, "response body is null").source().inputStream(), segmentUri);
```
Done.

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/SpooledSegment.java:
```diff
@@ -0,0 +1,53 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SpooledSegment
```
Ideally Segments would be sealed but client modules are JDK8 :(

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/InlineSegment.java:
```diff
@@ -0,0 +1,51 @@
+
+import static java.lang.String.format;
+
+public class InlineSegment
```
Done

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/encoding/DecryptingQueryDataDecoder.java:
```diff
@@ -0,0 +1,106 @@
+            cipher.init(DECRYPT_MODE, key);
+            return cipher;
+        }
+        catch (Exception e) {
```
Done

wendigo commented on testing/trino-tests/src/test/java/io/trino/tests/TestServer.java:
```diff
@@ -439,4 +450,19 @@ protected JsonResponse<QueryResults> computeNext(JsonResponse<QueryResults> prev
+
+    public static JsonCodec<QueryResults> queryResultsCodec()
+    {
+        Injector injector = new Bootstrap(
```
Done

wendigo commented on core/trino-main/pom.xml:
```diff
@@ -18,6 +18,7 @@
     </properties>
 
     <dependencies>
+
```
This is used and needed. OkHttp is used in the airlift opentelemetry tracing exporter (thus these dependencies as transitive ones, in order to be in the test scope, need to be declared in the runtime scope)

wendigo commented on core/trino-main/pom.xml:
```diff
@@ -398,6 +399,12 @@
             <scope>runtime</scope>
         </dependency>
 
+        <dependency>
```
It is - transitively:

```
[ERROR] Dependency chain:
[ERROR] io.trino:trino-main:jar:456-SNAPSHOT
[ERROR] \- io.airlift:tracing:jar:262:compile
[ERROR]    \- io.opentelemetry:opentelemetry-exporter-otlp:jar:1.41.0:compile
[ERROR]       \- io.opentelemetry:opentelemetry-exporter-sender-okhttp:jar:1.41.0:runtime
[ERROR]          \- com.squareup.okhttp3:okhttp:jar:4.12.0:runtime
```

electrum replied:
Makes sense, but why is `okhttp-urlconnection` need now, when it wasn't needed before?

wendigo replied:
I'll investigate it

wendigo replied:
Seems like it was some leftover and this is no longer needed.

wendigo commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/SwitchingFileSystem.java:
```diff
@@ -0,0 +1,152 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.groupingBy;
+
+final class SwitchingFileSystem
```
Done

wendigo commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpooledSegmentHandle.java:
```diff
@@ -0,0 +1,43 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+
+public record FileSystemSpooledSegmentHandle(String name, Date validUntil)
```
Done

wendigo commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java:
```diff
@@ -0,0 +1,136 @@
+            output.writeBytes(utf8Slice(fileHandle.name()));
+            return output.slice();
+        }
+        catch (IOException e) {
```
Then this method will throw IOException

wendigo replied:
ok, when try-with-resource is removed then this won't be thrown, done

wendigo commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java:
```diff
@@ -0,0 +1,136 @@
+    @Override
+    public SpooledSegmentHandle deserialize(Slice slice)
+    {
+        try (SliceInput input = slice.getInput()) {
```
Done.

wendigo commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManagerFactory.java:
```diff
@@ -0,0 +1,62 @@
+    {
+        requireNonNull(config, "requiredConfig is null");
+
+        // A plugin is not required to use Guice; it is just very convenient
```
Done

wendigo commented on core/trino-main/src/test/java/io/trino/server/protocol/TestQueryResultsSerialization.java:
```diff
@@ -0,0 +1,188 @@
+    private static final List<Column> COLUMNS = ImmutableList.of(new Column("_col0", BIGINT, new ClientTypeSignature("bigint")));
+
+    // As close as possible to the server mapper (client mapper differs)
+    private static final ObjectMapper MAPPER = new ObjectMapperProvider().get()
+            .registerModules(new QueryDataJacksonModule(), new QueryDataClientJacksonModule());
```
I've used two separate codecs instead to test how the client-server interacts:

```
    // As close as possible to the server mapper (client mapper differs)
    private static final io.airlift.json.JsonCodec<QueryResults> SERVER_CODEC = new JsonCodecFactory(new ObjectMapperProvider()
            .withModules(Set.of(new QueryDataJacksonModule())))
            .jsonCodec(QueryResults.class);

    private static final JsonCodec<QueryResults> CLIENT_CODEC = jsonCodec(QueryResults.class);
```

wendigo commented on core/trino-main/src/test/java/io/trino/server/protocol/TestQueryResultsSerialization.java:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+    private static String removeNewLines(String value)
```
Not needed. Dropped

wendigo commented on core/trino-main/src/test/java/io/trino/server/TestQueryStateInfoResource.java:
```diff
@@ -243,4 +253,18 @@ public void testGetQueryStateInfoNo()
+
+    public static JsonCodec<QueryResults> queryResultsCodec()
+    {
+        Injector injector = new Bootstrap(
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/sql/planner/sanity/ValidateDependenciesChecker.java:
```diff
@@ -497,7 +499,7 @@ public Void visitOutput(OutputNode node, Set<Symbol> boundSymbols)
             source.accept(this, boundSymbols); // visit child
 
-            checkDependencies(source.getOutputSymbols(), node.getOutputSymbols(), "Invalid node. Output column dependencies (%s) not in source plan output (%s)", node.getOutputSymbols(), source.getOutputSymbols());
+            checkDependencies(source.getOutputSymbols(), filter(node.getOutputSymbols(), symbol -> !symbol.equals(SPOOLING_METADATA_SYMBOL)), "Invalid node. Output column dependencies (%s) not in source plan output (%s)", node.getOutputSymbols(), source.getOutputSymbols());
```
Because the operator outputs an union (additional symbol to represent spooled handles)

wendigo commented on core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:
```diff
@@ -533,6 +536,7 @@ private static String formatFragment(
         PartitioningScheme partitioningScheme = fragment.getOutputPartitioningScheme();
         List<String> layout = partitioningScheme.getOutputLayout().stream()
                 .map(anonymizer::anonymize)
+                .filter(value -> !value.equals(SPOOLING_METADATA_COLUMN_NAME))
```
Yes. This is an implementation detail that user shouldn't see

wendigo commented on core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -959,7 +971,20 @@ public PhysicalOperation visitExplainAnalyze(ExplainAnalyzeNode node, LocalExecu
+                    .flatMap(encoderSelector::select);
+            PhysicalOperation operation = node.getSource().accept(this, context);
+
+            if (encoderFactory.isEmpty() || !spoolingManager.isLoaded()) {
```
yeah, that would make sense

wendigo replied:
Done

wendigo commented on core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -959,7 +971,20 @@ public PhysicalOperation visitExplainAnalyze(ExplainAnalyzeNode node, LocalExecu
+
+            Map<Symbol, Integer> spooledLayout = OutputSpoolingOperatorFactory.spooledLayout(operation.layout);
+            List<OutputColumn> outputLayout = spooledOutputLayout(node, spooledLayout);
+            return new PhysicalOperation(new OutputSpoolingOperatorFactory(context.getNextOperatorId(), node.getId(), spooledLayout, encoderFactory.orElseThrow().create(session, outputLayout), spoolingManager), spooledLayout, operation);
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/ZstdQueryDataEncoder.java:
```diff
@@ -0,0 +1,48 @@
+public class ZstdQueryDataEncoder
+        extends CompressedQueryDataEncoder
+{
+    private static final int COMPRESSION_THRESHOLD = 2_048;
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/JsonQueryDataEncoder.java:
```diff
@@ -0,0 +1,174 @@
+        }
+    }
+
+    public static class SnappyFactory
```
Ok, Snappy dropped

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/JsonQueryDataEncoder.java:
```diff
@@ -0,0 +1,174 @@
+public class JsonQueryDataEncoder
+        implements QueryDataEncoder
+{
+    private static final String ENCODING_ID = "json-ext";
```
`-ext` dropped

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/JsonQueryDataEncoder.java:
```diff
@@ -0,0 +1,174 @@
+            mapper.writeValue(wrapper, values);
+            List<Throwable> exceptions = serializationExceptions.build();
+            if (!exceptions.isEmpty()) {
+                throw new RuntimeException("Could not serialize to JSON", exceptions.getFirst());
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/EncryptingQueryDataEncoder.java:
```diff
@@ -0,0 +1,121 @@
+            cipher.init(ENCRYPT_MODE, privateKey);
+            return cipher;
+        }
+        catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException e) {
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/EncryptingQueryDataEncoder.java:
```diff
@@ -0,0 +1,121 @@
+public class EncryptingQueryDataEncoder
+        implements QueryDataEncoder
+{
+    private static final String CIPHER_NAME = "AES";
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/Session.java:
```diff
@@ -628,6 +642,16 @@ public SecurityContext toSecurityContext()
         return new SecurityContext(getRequiredTransactionId(), getIdentity(), queryId, start);
     }
 
+    public Optional<String> getQueryDataEncodingId()
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/cost/PlanNodeStatsEstimate.java:
```diff
@@ -82,6 +83,7 @@ public double getOutputSizeInBytes(Collection<Symbol> outputSymbols)
         requireNonNull(outputSymbols, "outputSymbols is null");
 
         return outputSymbols.stream()
+                .filter(symbol -> !symbol.type().equals(SPOOLING_METADATA_TYPE)) // We don't extra metadata block stats to be accounted for
```
Done

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/encoding/SnappyQueryDataDecoder.java:
```diff
@@ -0,0 +1,50 @@
+        SnappyDecompressor decompressor = new SnappyDecompressor();
+        byte[] bytes = ByteStreams.toByteArray(stream);
+        byte[] output = new byte[uncompressedSize];
+        verify(decompressor.decompress(bytes, 0, bytes.length, output, 0, output.length) == uncompressedSize, "Decompressed size does not match expected size");
```
Done.

wendigo replied:
Actually Snappy was removed (as its' server counterpart)

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/encoding/Lz4QueryDataDecoder.java:
```diff
@@ -0,0 +1,50 @@
+        Lz4Decompressor decompressor = new Lz4Decompressor();
+        byte[] bytes = ByteStreams.toByteArray(stream);
+        byte[] output = new byte[uncompressedSize];
+        verify(decompressor.decompress(bytes, 0, bytes.length, output, 0, output.length) == uncompressedSize, "Decompressed size does not match expected size");
```
Done.

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/encoding/QueryDataDecoders.java:
```diff
@@ -0,0 +1,55 @@
+
+public class QueryDataDecoders
+{
+    private static Map<String, QueryDataDecoder.Factory> factories = new HashMap<>();
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/CompressedQueryDataEncoder.java:
```diff
@@ -0,0 +1,71 @@
+
+    protected static int pagesSize(List<Page> pages)
+    {
+        return saturatedCast(pages.stream()
+                .map(Page::getSizeInBytes)
+                .reduce(0L, Long::sum));
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingServerModule.java:
```diff
@@ -0,0 +1,61 @@
+        SpoolingConfig spoolingConfig = buildConfigObject(SpoolingConfig.class);
+        boolean isCoordinator = buildConfigObject(ServerConfig.class).isCoordinator();
+        if (spoolingConfig.isEnabled()) {
+            configBinder(binder).bindConfig(SpoolingConfig.class);
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingServerModule.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    protected void setup(Binder binder)
+    {
+        SpoolingConfig spoolingConfig = buildConfigObject(SpoolingConfig.class);
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingManagerBridge.java:
```diff
@@ -0,0 +1,163 @@
+    private Slice encrypt(Slice input)
+    {
+        try {
+            Cipher cipher = Cipher.getInstance("AES");
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingManagerBridge.java:
```diff
@@ -0,0 +1,163 @@
+    private Slice decrypt(Slice input)
+    {
+        try {
+            Cipher cipher = Cipher.getInstance("AES");
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingManagerBridge.java:
```diff
@@ -0,0 +1,163 @@
+    {
+        try {
+            Cipher cipher = Cipher.getInstance("AES");
+            cipher.init(ENCRYPT_MODE, secretKey);
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingConfig.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @ConfigDescription("256 bit, base64-encoded secret key used to secure segment identifiers")
+    @Config("protocol.spooling.encryption-key")
```
makes sense. renamed

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SegmentResource.java:
```diff
@@ -0,0 +1,116 @@
+    @GET
+    @Produces(MediaType.APPLICATION_OCTET_STREAM)
+    @ResourceSecurity(PUBLIC)
+    @Path("")
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SegmentResource.java:
```diff
@@ -0,0 +1,116 @@
+    {
+        return UriBuilder.fromUri(info.baseUriBuilder().build())
+                .path(SegmentResource.class)
+                .path(SegmentResource.class, "download");
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/JsonQueryDataEncoder.java:
```diff
@@ -0,0 +1,174 @@
+    public DataAttributes encodeTo(OutputStream output, List<Page> pages)
+            throws IOException
+    {
+        ImmutableList.Builder<Throwable> serializationExceptions = ImmutableList.builder();
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SegmentResource.java:
```diff
@@ -0,0 +1,116 @@
+    public HostAddress nextActiveNode()
+    {
+        List<InternalNode> internalNodes = ImmutableList.copyOf(nodeManager.getActiveNodesSnapshot().getAllNodes());
+        return internalNodes.get(toIntExact(nextWorkerIndex.incrementAndGet() % internalNodes.size()))
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/QueryDataProducer.java:
```diff
@@ -0,0 +1,113 @@
+    private final Optional<QueryDataEncoder.Factory> encoderFactory;
+    private long currentOffset;
+
+    private AtomicBoolean metadataWritten = new AtomicBoolean(false);
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/QueryDataProducer.java:
```diff
@@ -0,0 +1,113 @@
+                }
+            }
+        }
+        catch (Exception e) {
```
Not applied. After changes this code looks differently.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/QueryDataJacksonModule.java:
```diff
@@ -0,0 +1,114 @@
+        public void serialize(Segment value, JsonGenerator gen, SerializerProvider provider)
+                throws IOException
+        {
+            createSerializer(provider, provider.constructSpecializedType(provider.constructType(Segment.class), value.getClass())).serializeWithType(value, gen, provider, segmentSerializer(provider));
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:
```diff
@@ -0,0 +1,435 @@
+            SpoolingContext spoolingContext = new SpoolingContext(operatorContext.getDriverContext().getSession().getQueryId(), rows);
+            SpooledSegmentHandle segmentHandle = spoolingManager.create(spoolingContext);
+
+            try (OutputStream output = spoolingManager.createOutputStream(segmentHandle); ByteArrayOutputStream bufferedOutput = new ByteArrayOutputStream(toIntExact(controller.getCurrentSpooledSegmentTarget()))) {
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:
```diff
@@ -0,0 +1,435 @@
+        }
+    }
+
+    private record PageBuffer(LinkedList<Page> buffer)
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:
```diff
@@ -0,0 +1,435 @@
+
+        public static PageBuffer create()
+        {
+            return new PageBuffer(new LinkedList<>());
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:
```diff
@@ -0,0 +1,435 @@
+        }
+    }
+
+    private record PageBuffer(LinkedList<Page> buffer)
```
Done

wendigo commented on core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:
```diff
@@ -0,0 +1,435 @@
+        public OutputSpoolingInfo mergeWith(OutputSpoolingInfo other)
+        {
+            return new OutputSpoolingInfo(
+                    new Duration(encodingWallTime.toMillis() + other.encodingWallTime().toMillis(), MILLISECONDS).convertToMostSuccinctTimeUnit(),
```
Done

wendigo commented on core/trino-main/src/test/java/io/trino/server/protocol/TestQueryResultsSerialization.java:
```diff
@@ -0,0 +1,188 @@
+
+    private static String removeNewLines(String value)
+    {
+        return value.replaceAll("\n", "").replaceAll("\r", "").replaceAll("\\s+", "");
```
I've dropped this code altogether

wendigo commented:
Question regarding encryption @electrum:

for AES/GCM we need to send IV to client as well. Is the separate data attribute ok or should we write in our own framing like this:

`IV_length long | IV byte[length] | encoded message byte[]`

wendigo commented:
@electrum PTAL

wendigo commented:
AC @electrum

wendigo commented:
Green again :)

electrum reviewed (commented):
Aside from encryption and some questions about the configs, I think this is ready to ship!

wendigo commented on core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java:
```diff
@@ -285,9 +291,11 @@ private Page emptySingleRowPage(Map<Symbol, Integer> layout, Block block)
-            return page.stream().map(reduce).reduce(0L, Long::sum);
+            return page.stream()
+                    .mapToLong(reduce)
+                    .reduce(0L, Long::sum);
```
Under the hood this is the same:

```
@Override
    public final long sum() {
        // use better algorithm to compensate for intermediate overflow?
        return reduce(0, Long::sum);
    }
```

wendigo commented on core/trino-main/pom.xml:
```diff
@@ -351,6 +356,12 @@
             <version>9.11.1</version>
         </dependency>
 
+        <dependency>
```
I've moved to tink - it's a little bit opinionated but will definitely streamline the support in other clients

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingServerModule.java:
```diff
@@ -31,14 +39,28 @@ public class SpoolingServerModule
 
-        boolean isCoordinator = buildConfigObject(ServerConfig.class).isCoordinator();
+        // Once spooling is enabled we can switch query data encryption on by default
+        configBinder(binder).bindConfigDefaults(FeaturesConfig.class,
```
Done.

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingManagerBridge.java:
```diff
@@ -54,7 +54,7 @@ public SpoolingManagerBridge(SpoolingConfig spoolingConfig, SpoolingManagerRegis
         this.maximumSegmentSize = spoolingConfig.getMaximumSegmentSize();
         this.inlineSegments = spoolingConfig.isInlineSegments();
-        this.secretKey = spoolingConfig.getEncryptionKey().orElse(createRandomAesEncryptionKey());
+        this.secretKey = spoolingConfig.getSharedEncryptionKey().orElse(createRandomAesEncryptionKey());
```
It's actually verified in the config that the shared key was set. This is never reached but changed that to orElseThrow()

wendigo commented:
Reworked encryption to be server-side rather than client-side. This makes a lot of things easier and design more robust.

I've updated design doc to reflect these changes.

electrum commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingConfig.java:
```diff
@@ -0,0 +1,136 @@
+    {
+        return sharedEncryptionKey
+                .map(Ciphers::is256BitSecretKeySpec)
+                .orElse(false);
```
This should be `true` so that we only fail when this property is set. The method below fails when not set.

electrum commented on core/trino-spi/src/main/java/io/trino/spi/protocol/SpoolingManager.java:
```diff
@@ -0,0 +1,59 @@
+        throw new UnsupportedOperationException();
+    }
+
+    default Map<String, String> headers(SpooledSegmentHandle deserialize)
```
I see why you made this separate from `directLocation()`. Let's move it directly afterwards, since they logically go together.

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/EncryptionUtils.java:
```diff
@@ -0,0 +1,113 @@
+            CleartextKeysetHandle.write(aes256Keyset, withOutputStream(outputStream));
+            return wrappedBuffer(outputStream.toByteArray());
+        }
+        catch (Exception e) {
```
```suggestion
        catch (IOException | GeneralSecurityException e) {
```

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/EncryptionUtils.java:
```diff
@@ -0,0 +1,113 @@
+        try {
+            return CleartextKeysetHandle.read(BinaryKeysetReader.withBytes(key.getBytes()));
+        }
+        catch (Exception e) {
```
```suggestion
        catch (IOException | GeneralSecurityException e) {
```

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/ExceptionMappingInputStream.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Override
+    public void mark(int readlimit)
```
No need to override

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/ExceptionMappingInputStream.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Override
+    public boolean markSupported()
```
No need to override

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/ExceptionMappingInputStream.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Override
+    public void reset()
```
This shouldn't be needed as Trino file system doesn't implement it

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/ExceptionMappingInputStream.java:
```diff
@@ -0,0 +1,124 @@
+            return super.read();
+        }
+        catch (FileNotFoundException e) {
+            throw new RuntimeException("Segment not found or expired", e);
```
These should rethrow `IOException`, as it's strange to throw an unchecked exception here

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java:
```diff
@@ -0,0 +1,179 @@
+        checkExpiration(segmentHandle);
+        try {
+            if (!fileSystem.newInputFile(segmentLocation(segmentHandle)).exists()) {
+                throw new RuntimeException("Segment not found or expired");
```
`IOException`

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java:
```diff
@@ -0,0 +1,179 @@
+                    return decryptingInputStream(stream, key);
+                }
+                catch (IOException e) {
+                    throw new RuntimeException(e);
```
We shouldn't rethrow here, as `openInputStream()` is expected to throw `IOException`

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java:
```diff
@@ -0,0 +1,179 @@
+            }).orElse(stream));
+        }
+        catch (FileNotFoundException e) {
+            throw new RuntimeException("Segment not found or expired", e);
```
`IOException`

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FilesystemSpoolingModule.java:
```diff
@@ -0,0 +1,77 @@
+        if (config.isNativeS3Enabled()) {
+            install(new S3FileSystemModule());
+            factories.addBinding("s3").to(S3FileSystemFactory.class);
+            factories.addBinding("s3a").to(S3FileSystemFactory.class);
+            factories.addBinding("s3n").to(S3FileSystemFactory.class);
```
We don't need thse

electrum commented on plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FilesystemSpoolingModule.java:
```diff
@@ -0,0 +1,77 @@
+        if (config.isNativeAzureEnabled()) {
+            install(new AzureFileSystemModule());
+            factories.addBinding("abfs").to(AzureFileSystemFactory.class);
+            factories.addBinding("abfss").to(AzureFileSystemFactory.class);
```
We don't need this

wendigo commented on client/trino-client/src/main/java/io/trino/client/spooling/SegmentLoader.java:
```diff
@@ -0,0 +1,111 @@
+            public void close()
+                    throws IOException
+            {
+                response.close();
```
JDK8 - reference is not enough. Changed to:

```
try (Response ignored = response; InputStream ignored2 = delegate) {
                    delete(segmentUri);
                }
```

wendigo commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/SpoolingManagerBridge.java:
```diff
@@ -0,0 +1,166 @@
+        this.initialSegmentSize = spoolingConfig.getInitialSegmentSize();
+        this.maximumSegmentSize = spoolingConfig.getMaximumSegmentSize();
+        this.inlineSegments = spoolingConfig.isInlineSegments();
+        this.secretKey = spoolingConfig.getSharedEncryptionKey().orElseThrow();
```
No, handles are returned to the client as a identifier used to download the segment, hence the encryption.

electrum commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/JsonQueryDataEncoder.java:
```diff
@@ -69,9 +90,52 @@ public DataAttributes encodeTo(OutputStream output, List<Page> pages)
+                }
+                generator.writeEndObject();
+            }
+            default -> throw new UnsupportedOperationException("Unsupported type: " + value.getClass().getName());
```
This will fail on types with custom serialization, e.g. `BingTile`

electrum replied:
We should be able to handle this with a fallback:
```java
default -> mapper.writeValue(generator, value);

electrum commented on core/trino-main/src/main/java/io/trino/server/protocol/spooling/encoding/JsonQueryDataEncoder.java:
```diff
@@ -127,7 +133,7 @@ private static void writeValue(JsonGenerator generator, Object value)
                 generator.writeEndObject();
             }
-            default -> throw new UnsupportedOperationException("Unsupported type: " + value.getClass().getName());
+            default -> generator.writeString(value.toString());
```
This needs to use `ObjectMapper` for types like `BingTile` that use custom serialization.

wendigo replied:
Done. Makes sense.

wendigo commented on client/trino-client/src/main/java/io/trino/client/QueryResults.java:
```diff
@@ -138,10 +111,15 @@ public List<Column> getColumns()
+    }
+
+    @JsonProperty("data")
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
```
It's actually needed. Some clients like ODBC assume that `"data": null` won't be ever emitted, in that case this annotation does exactly that - hides the null data field

wendigo commented:
@electrum ptal

electrum reviewed (approved):
Let's ship this. Remember to squash all the commits.

mosabua commented:
No release notes entry @electrum @wendigo @nineinchnick ?

wendigo commented:
@mosabua correct. This is an experimental feature with no client to use it so definitely not a feature
